relay-sampling/src/lib.rs

```diff
@@ -115,6 +116,51 @@ impl EqCondition {
     }
 }
 
+macro_rules! impl_cmp_condition {
+    ($struct_name:ident, $operator:tt) => {
+        #[derive(Debug, Clone, Serialize, Deserialize)]
+        pub struct $struct_name {
+            pub name: String,
+            pub value: Number,
+        }
+
+        impl $struct_name {
+            fn matches_event(&self, event: &Event) -> bool {
+                self.matches(event)
+            }
+            fn matches_trace(&self, trace: &TraceContext) -> bool {
+                self.matches(trace)
+            }
+
+            fn matches<T: FieldValueProvider>(&self, value_provider: &T) -> bool {
+                let value = match value_provider.get_value(self.name.as_str()) {
+                    Value::Number(x) => x,
+                    _ => return false
+                };
+
+                // Try various conversion functions in order of expensiveness and likelihood
+                // - as_i64 is not really fast, but most values in sampling rules can be i64, so we could
+                //   return early
+                // - f64 is more likely to succeed than u64, but we might lose precision
+                if let (Some(a), Some(b)) = (value.as_i64(), self.value.as_i64()) {
+                    a $operator b
+                } else if let (Some(a), Some(b)) = (value.as_u64(), self.value.as_u64()) {
+                    a $operator b
+                } else if let (Some(a), Some(b)) = (value.as_f64(), self.value.as_f64()) {
+                    a $operator b
+                } else {
+                    false
+                }
+            }
+        }
+    }
+}
+
+impl_cmp_condition!(GteCondition, >=);
+impl_cmp_condition!(LteCondition, <=);
+impl_cmp_condition!(LtCondition, <);
+impl_cmp_condition!(GtCondition, >);
```

Comment:
Out of curiosity, would there be a way to do this without a macro? Something like `type GtCondition = CmpCondition<std::cmp::PartialOrd::gt>`. I guess you would run into problems because you have three different operand types `i64`, `u64`, `f64` inside?

Reply:
I didn't find a way, you'd have to be generic over a type, and functions are either:

* a singular type `fn(a: i64, b: i64) -> bool` (regular functions)
* an unnameable type (closures)

there's no const generics for function pointers yet

In order to make generics truly work, you'd have to define a trait `Comparator`, implement it for a few zero-sized types `Lt, Gt, ...`:

```
trait Comparator {
    fn matches(a: Number, b: Number) -> bool;
}

struct Lt;
impl Comparator for Lt { ... }

struct CmpCondition<A> { ... }
impl<A: Comparator> CmpCondition<A> {
    fn matches(self, event: ...) -> bool {
        A::matches(self.number, get_event_number(event))
    }
}

type LtCondition = CmpCondition<Lt>;
```

but this seemed like more effort

I think instead of being generic over a type what you could do is to write a single non-generic type:

```
struct CmpCondition {
    cmp_fn: fn(a: Number, b: Number) -> bool;
    ...
}
```

but then you have to write custom deserialization logic, and serialization just won't work unless you keep the enum variants:

```
enum RuleCondition {
    Gte(CmpCondition),
    Lt(CmpCondition)
}
```

Reply:
Thanks for the rundown!